Celeste.Mod.mm/Patches/Monocle/Tracker.cs

```diff
@@ -123,5 +123,109 @@ private static List<Type> GetSubclasses(Type type) {
+                    TrackedComponentTypes[type] = value.Distinct().ToList();
+                }
+                foreach(Type subtype in subtypes) {
+                    if (!subtype.IsAbstract) {
+                        if (!TrackedComponentTypes.TryGetValue(subtype, out List<Type> value)) {
+                            value = new List<Type>();
+                            TrackedComponentTypes.Add(subtype, value);
+                        }
+                        value.AddRange(TrackedComponentTypes.TryGetValue(trackedAs != null && trackedAs.IsAssignableFrom(type)
+                            ? trackedAs : type, out List<Type> list) ? list : new List<Type>());
+                        TrackedComponentTypes[subtype] = value.Distinct().ToList();
+                    }
+                }
+            }
+            else {
+                throw new Exception("Type '" + type.Name + "' cannot be TrackedAs because it does not derive from Entity or Component");
+            }
+            RefreshTracker(type);
+        }
+
+        public static void RefreshTracker(Type type) {
+            if (typeof(Entity).IsAssignableFrom(type) && !Engine.Scene.Tracker.Entities.ContainsKey(type)) {
+                Engine.Scene.Tracker.Entities.Add(type, new List<Entity>());
+            }
+            else if (typeof(Component).IsAssignableFrom(type) && !Engine.Scene.Tracker.Components.ContainsKey(type)) {
+                Engine.Scene.Tracker.Components.Add(type, new List<Component>());
+            } else {
+                throw new Exception("Type '" + type.Name + "' cannot be TrackedAs because it does not derive from Entity or Component");
+            }
+            RefreshTrackerLists();
+        }
+
+        public static void RefreshTracker() {
+            foreach (Type entityType in StoredEntityTypes) {
+                if (!Engine.Scene.Tracker.Entities.ContainsKey(entityType)) {
+                    Engine.Scene.Tracker.Entities.Add(entityType, new List<Entity>());
+                }
+            }
+            foreach (Type componentType in StoredComponentTypes) {
+                if (!Engine.Scene.Tracker.Components.ContainsKey(componentType)) {
+                    Engine.Scene.Tracker.Components.Add(componentType, new List<Component>());
+                }
+            }
+            RefreshTrackerLists();
+        }
+
+        private static void RefreshTrackerLists() {
+            foreach (Entity entity in Engine.Scene.Entities) {
+                foreach (Component component in entity.Components) {
+                    Type componentType = component.GetType();
+                    if (!TrackedComponentTypes.TryGetValue(componentType, out List<Type> componentTypes)) {
+                        continue;
+                    }
+                    foreach (Type trackedType in componentTypes) {
+                        Engine.Scene.Tracker.Components[trackedType].Add(component);
```

Comment:
Have you tested that this doesn't cause entities/components to get tracked twice?

Reply:
(Safest bet would be to call .Clear on both the Entities and Components dictionaries)

Reply:
I can do that